prusti-specs/src/common.rs

```diff
@@ -1,29 +1,358 @@
+                    let new_group =
+                        proc_macro2::Group::new(group.delimiter(), self.rewrite_tokens(group.stream()));
+                    TokenTree::Group(new_group)
+                }
+                TokenTree::Ident(ident) if ident == "self" => {
+                    TokenTree::Ident(proc_macro2::Ident::new("_self", ident.span()))
+                },
+                other => other,
+            }));
+            parse_quote_spanned! {tokens_span=>
+                #rewritten
+            }
+        }
+    }
+
+    impl<'a> VisitMut for Rewriter<'a> {
+        fn visit_fn_arg_mut(&mut self, fn_arg: &mut FnArg) {
+            if let FnArg::Receiver(receiver) = fn_arg {
+                let span = receiver.span();
+                let and = if receiver.reference.is_some() {
+                    // TODO: do lifetimes need to be specified here?
+                    quote_spanned! {span=> &}
+                } else {
+                    quote! {}
+                };
+                let mutability = &receiver.mutability;
+                let new_ty = self.new_ty;
+                let new_fn_arg: FnArg = parse_quote_spanned! {span=>
+                _self : #and #mutability #new_ty
+            };
+                *fn_arg = new_fn_arg;
+            } else {
+                syn::visit_mut::visit_fn_arg_mut(self, fn_arg);
+            }
+        }
+
+        fn visit_ident_mut(&mut self, ident: &mut Ident) {
+            if ident == "self" {
+                *ident = Ident::new("_self", ident.span());
+            }
+            syn::visit_mut::visit_ident_mut(self, ident)
+        }
+
+        fn visit_macro_mut(&mut self, makro: &mut Macro) {
+            // A macro can appear in a spec function (e.g. `matches!`)
+            makro.tokens = self.rewrite_tokens(makro.tokens.clone());
+            syn::visit_mut::visit_macro_mut(self, makro);
+        }
     }
 }
 
-impl HasGenerics for syn::ItemImpl {
-    fn get_generics(&self) -> &syn::Generics {
-        &self.generics
+/// Copies the [syn::Generics] of `source` to the generics of `target`
+/// **Important**: Lifetimes and const params are currently ignored
+pub(crate) fn merge_generics<T: HasGenerics>(target: &mut T, source: &T) {
+    let generics_target = target.generics_mut();
+    let generics_source = source.generics();
+
+    // This is not particularly fast or "elegant", but we can expect a reasonably
+    // small amount of type parameters in general (n < 10)
+    for param_source in generics_source.params.iter() {
+        // Lifetimes and consts are currently not handled
+        if let GenericParam::Type(type_param_source) = param_source {
+            let in_target = generics_target.params.iter_mut()
+                .find(|gp| match gp {
+                    GenericParam::Type(tp) => tp.ident.eq(&type_param_source.ident),
+                    _ => false
+                }
+                ).map(|gp| force_matches!(gp, GenericParam::Type(tp) => tp));
```

Comment:
This sounds like you want a `HashSet` of the identifiers already inserted...

Reply:
It screams for `HashABC`s ^^
I refactored it again. Maybe it could be done more elegantly?